source/shared/core_conn.cpp

```diff
@@ -150,21 +150,55 @@ sqlsrv_conn* core_sqlsrv_connect( _In_ sqlsrv_context& henv_cp, _In_ sqlsrv_cont
     
     build_connection_string_and_set_conn_attr( conn, server, uid, pwd, options_ht, valid_conn_opts, driver, conn_str TSRMLS_CC );
     
-    bool is_missing_driver = false;
+    bool is_missing_driver = true;
 
     if ( conn->is_driver_set ) {
         r = core_odbc_connect( conn, conn_str, is_missing_driver, is_pooled );
     }
     else if ( conn->ce_option.enabled ) {
+#ifndef _WIN32
+        // In non-Windows environment, because different unixODBC versions return different error states,
+        // if it fails to connect it is unreliable to check for a certain sql state error. Thus, 
+        // we will ignore the returned value of 'is_missing_driver' and simply throw an exception.
+        if( core_search_odbc_driver( DRIVER_VERSION::ODBC_DRIVER_17 ) ) {
+            conn_str = conn_str + CONNECTION_STRING_DRIVER_NAME[ DRIVER_VERSION::ODBC_DRIVER_17 ];
+            r = core_odbc_connect( conn, conn_str, is_missing_driver, is_pooled );
+        }
+        else {
+            CHECK_CUSTOM_ERROR( true, conn, SQLSRV_ERROR_AE_DRIVER_NOT_INSTALLED, get_processor_arch()) {
+                throw core::CoreException();
+            }
+        }
+#else
         conn_str = conn_str + CONNECTION_STRING_DRIVER_NAME[ DRIVER_VERSION::ODBC_DRIVER_17 ];
         r = core_odbc_connect( conn, conn_str, is_missing_driver, is_pooled );
 
         CHECK_CUSTOM_ERROR( is_missing_driver, conn, SQLSRV_ERROR_AE_DRIVER_NOT_INSTALLED, get_processor_arch()) {
             throw core::CoreException();
         }
+#endif // !_WIN32 
     }
     else {
-
+#ifndef _WIN32
+        is_missing_driver = true;
+        for ( std::size_t i = DRIVER_VERSION::FIRST; i <= DRIVER_VERSION::LAST; ++i ) {
+            if (i == DRIVER_VERSION::ODBC_DRIVER_11)    // skip ODBC 11 in non-Windows environment
+                continue;
+
+            if( core_search_odbc_driver( i ) ) {
+                std::string conn_str_driver = conn_str + CONNECTION_STRING_DRIVER_NAME[ DRIVER_VERSION( i ) ];
+                r = core_odbc_connect( conn, conn_str_driver, is_missing_driver, is_pooled );
+                // simply check the connection succeeded or not and ignore the returned value of 'is_missing_driver' 
+                if ( SQL_SUCCEEDED( r ) ) {
+                    is_missing_driver = false;
+                    break;
+                }
+            }
+        }
+        CHECK_CUSTOM_ERROR( is_missing_driver, conn, SQLSRV_ERROR_DRIVER_NOT_INSTALLED, get_processor_arch()) {
```

Comment:
this should be check in line 215 (after `if ( conn->is_driver_set )`, `else if ( conn->ce_option.enabled )` and `else `blocks )